packages/flutter/lib/src/material/adaptive.dart

```diff
@@ -0,0 +1,58 @@
+// Copyright 2018 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/widgets.dart';
+
+import 'theme.dart';
+
+/// Options for the adaptive behavior of [AdaptiveWidget].
+enum AdaptiveMode {
+  /// Always adapt the widget to the current [TargetPlatform].
+  adaptive,
+  /// Never adapt the widget to the current [TargetPlatform]. Always return
+  /// the Material Design widget.
+  static,
+  /// Adapt the widget to the current [TargetPlatform] if the widget type is
+  /// enabled in [ThemeData.adaptiveWidgetTheme].
+  theme,
+}
+
+/// A [StatelessWidget] that can build different widgets according to the
+/// current platform environment depending on the [adaptiveMode].
+abstract class AdaptiveWidget extends StatelessWidget {
```

Comment:
rather than using inheritance here it's probably going to be more convenient to use composition. You could have a widget that takes two builders, one for material and one for cupertino.

Reply:
That would make it easier for people to autoadapt in their apps, without our help.
Also it would work better for widgets that already use inheritance, e.g. the material buttons.

Reply:
the widget could take a Type to look up in the map, which if null just means "always adapt, based on the theme's platform".

Reply:
Ya, totally. Taking builders makes sense.